tools/pre-push

```diff
@@ -0,0 +1,83 @@
+#!/bin/sh
+
+set -eu
+
+if [ -t 2 ]; then
+    red="\e[1;31m"
+    green="\e[1;32m"
+    blue="\e[1;34m"
+    grey="\e[0m"
+else
+    red=""
+    green=""
+    blue=""
+    grey=""
+fi
+
+# checks a single commit (ref name in $1) for validity
+check_commit() {
+    tmpdir="tmp/pre-push/${commit}"
+    rm -rf "tmp/pre-push/${commit}"
+
+    # unpack all of the changed files, plus the test-static-code from that commit
+    files="$(git show --format='' --name-only "${commit}" --) tools/test-static-code"
+    git archive --prefix="${tmpdir}/" "${commit}" -- ${files} | tar x
+
+    # check that node_modules always gets updated with package.json
+    if [ -e "${tmpdir}/package.json" -a ! -e "${tmpdir}/node_modules" ]; then
```

Comment:
I checked out your branch, in a local `tss` branch of mine, adjusted the hook:

    lrwxrwxrwx. 1 martin martin   20 Jan 13 08:59 pre-push -> ../../tools/pre-push

and tried to push to my fork with `git push my`.
```
❱❱❱ git push my 
Performing pre-push checks for refs/heads/tss:
 - a01793db3 tools/pre-push: introduce a proper git pre-push hook: pass
 - 2b91fa724 test-static-code: clean up python checks:
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 85: tools/node-modules: No such file or directory
 - 8736283a8 storage: clean up vdo python imports:
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 77: tools/node-modules: No such file or directory
 - 6850624cd check-bastion test: don't 'import *':
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 77: tools/node-modules: No such file or directory
 - e87ea3ef2 tools/escape-to-c: drop a superfluous f-string:
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 77: tools/node-modules: No such file or directory
 - 076521182 pkg/lib: add 'import sys' to inotify.py:
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 77: tools/node-modules: No such file or directory
 - cb3e3aac6 tools/test-static-code: major overhaul:
    not ok 5 /static-code/test-node-modules-freshness
    # tools/test-static-code: line 77: tools/node-modules: No such file or directory
error: failed to push some refs to 'github.com:martinpitt/cockpit.git'
```

`tools/node_modules` obviously exists in my tree, but it didn't get copied into tmp/pre-push/sha/. The top-level commit passed, so this might only work for future once we land this. But of course it'd still break testing of stable branches.

Honestly I don't like this copying/temp git tree magic very much. It is brittle, hard to debug, and imposes huge assumptions on test-static-code: it has to work on any random combinations of changed files that we imagine, plus it doesn't check the corresponding dist/ anyway. It also seems to work against your goal of making the hook faster. The goal of checking every commit *is* nice, of course. So not a veto, but if you'd cut that back I wouldn't complain at all :grin:

Reply:
> tools/node_modules obviously exists in my tree, but it didn't get copied into tmp/pre-push/sha/. The top-level commit passed, so this might only work for future once we land this. But of course it'd still break testing of stable branches.

Most of the changes in this branch are specifically about preparing `tools/test-static-code` to work with this new model.  It makes quite a lot of assumptions (simple stuff like "there will be at least one python file to test") which are broken under this model, and that brokenness is demonstrated by the failing commits here.

> Honestly I don't like this copying/temp git tree magic very much.

So the rationale is something like this:
 - every single commit should have passing tests, as far as we can reasonably manage (think: our future selves bisecting stuff)
 - doing this all out of the git object store is problematic for a lot of reasons.  Specifically: the tools we want to run expect to inspect files on the filesystem.  Also, I've checked, and in a lot of cases, the git version of things is actually slower than the filesystem version, presumably related to decompression and lack of caching.  We want to run this on the filesystem.
 - considering that we may be pushing something that's not the currently checked out branch, or might be pushing a branch when we have changes in the working tree, there's actually absolutely no guarantee that the thing we want to test exists anywhere on the filesystem at all.
 - the only solution i can think of is that we need to check out the thing we actually want to test.  That's the entire basis of this approach, and I think it's pretty reasonable.

Reply:
Right, as I said I agree to the goal, I just don't like the complexity of it. (Kinda similar to the jumpstart discussion). But as I said it's not a veto.